bpf/lib/encrypt.h

```diff
@@ -222,6 +223,159 @@ do_decrypt(struct __ctx_buff *ctx, __u16 proto)
+ * 4. Traffic is already ESP encrypted
+ */
+static __always_inline int
+ipsec_redirect_sec_id_ok(__u32 src_sec_id, __u32 dst_sec_id, int ip_proto) {
+	if (ip_proto == IPPROTO_ESP)
+		return 0;
+	if (src_sec_id == HOST_ID)
+		return 0;
+	if (dst_sec_id == HOST_ID)
+		return 0;
+	if (!identity_is_cluster(dst_sec_id))
+		return 0;
+	if (!identity_is_cluster(src_sec_id))
+		return 0;
+	if (identity_is_remote_node(dst_sec_id))
+		return 0;
+	if (identity_is_remote_node(src_sec_id))
+		return 0;
+	return 1;
+}
+
+static __always_inline int
+ipsec_maybe_redirect_to_encrypt(struct __ctx_buff *ctx, __be16 proto,
+				__u32 src_sec_identity)
+{
+	struct remote_endpoint_info __maybe_unused *dst = NULL;
+	struct remote_endpoint_info __maybe_unused *src = NULL;
+	void *data __maybe_unused, *data_end __maybe_unused;
+	struct iphdr __maybe_unused *ip4;
+	struct ipv6hdr __maybe_unused *ip6;
+	__u32 magic __maybe_unused = 0;
+	int ip_proto = 0;
+	int ret = 0;
+	union macaddr dst_mac = CILIUM_NET_MAC;
+
+	if (!eth_is_supported_ethertype(proto))
+		return DROP_UNSUPPORTED_L2;
+
+	switch (proto) {
+# ifdef ENABLE_IPV4
+	case bpf_htons(ETH_P_IP):
+		if (!revalidate_data(ctx, &data, &data_end, &ip4))
+			return DROP_INVALID;
+
+#  if defined(TUNNEL_MODE)
+		/* tunnel mode needs a bit of special handling when
+		 * encapsulated packets get here the destination address is
+		 * already a cluster node IP.
+		 *
+		 * the security ID is appended to the mark in the overlay prog
+		 * and we can extract this with 'get_identity'.
+		 * additionally, this is a VXLAN packet so ip4->daddr is the ip
+		 * of the destination host already and can be passed into
+		 * set_ipsec_encrypt to obtain the correct node ID and spi.
+		 */
+		if (ctx_is_overlay(ctx)) {
+			/* if this is already encrypted, just pass it to stack */
+			if (ctx_is_overlay_encrypted(ctx))
+				return CTX_ACT_OK;
+			/* see comment in the native-routing mode call. */
+			ret = set_ipsec_encrypt(ctx, 0, ip4->daddr,
+						get_identity(ctx), true,
+						true);
+			if (ret != CTX_ACT_OK)
+				return ret;
+			goto overlay_encrypt;
+		}
+#  endif /* TUNNEL_MODE */
+
+		ip_proto = ip4->protocol;
+
+		dst = lookup_ip4_remote_endpoint(ip4->daddr, 0);
+
+		if (src_sec_identity == UNKNOWN_ID) {
+			src = lookup_ip4_remote_endpoint(ip4->saddr, 0);
+			if (!src)
+				return CTX_ACT_OK;
+
+			src_sec_identity = src->sec_identity;
+		}
+		break;
+# endif /* ENABLE_IPV4 */
+
+# ifdef ENABLE_IPV6
+	case bpf_htons(ETH_P_IPV6):
+#ifndef TUNNEL_MODE
+		/* handle native routing ipv6 */
+		if (!revalidate_data(ctx, &data, &data_end, &ip6))
+			return DROP_INVALID;
+
+		ip_proto = ip6->nexthdr;
+
+		dst = lookup_ip6_remote_endpoint((union v6addr *)&ip6->daddr, 0);
+
+		if (src_sec_identity == UNKNOWN_ID) {
+			src = lookup_ip6_remote_endpoint((union v6addr *)&ip6->saddr, 0);
+			if (!src)
+				return CTX_ACT_OK;
+
+			src_sec_identity = src->sec_identity;
+		}
+		break;
+#endif /* TUNNEL_MODE */
+# endif /* ENABLE_IPv6 */
```

Comment:
I'm not sure I understand why we need this IPv6 + tunneling case. We don't support IPv6 as an underlay, so in tunneling mode you shouldn't see any pod-to-pod IPv6 traffic passing here. It should all be in the IPv4 case. Of course, you can have non-VXLAN IPv6 traffic, but it won't be pod-to-pod.

Reply:
The macro is an `ifndef` not `ifdef`. 
We encrypt IPv6 traffic in native routing mode, right?

Reply:
:facepalm: Missed the `n`. Ok, so makes complete sense :sweat_smile: